Mathlib/Algebra/Group/WithOne/Defs.lean

```diff
@@ -0,0 +1,431 @@
+
+/-- Deconstruct a `x : with_one α` to the underlying value in `α`, given a proof that `x ≠ 1`. -/
+@[to_additive unzero
+      "Deconstruct a `x : with_zero α` to the underlying value in `α`, given a proof that `x ≠ 0`."]
+def unone {x : WithOne α} (hx : x ≠ 1) : α :=
+  WithBot.unbot x hx
+#align with_one.unone WithOne.unone
+
+@[simp, to_additive unzero_coe]
+theorem unone_coe {x : α} (hx : (x : WithOne α) ≠ 1) : unone hx = x :=
+  rfl
+#align with_one.unone_coe WithOne.unone_coe
+
+@[simp, to_additive coe_unzero]
+theorem coe_unone {x : WithOne α} (hx : x ≠ 1) : ↑(unone hx) = x :=
+  WithBot.coe_unbot x hx
+#align with_one.coe_unone WithOne.coe_unone
+
+-- porting note: in Lean 4 this is a syntactic tautology
+-- @[to_additive]
+-- theorem some_eq_coe {a : α} : (some a : WithOne α) = ↑a :=
+--   rfl
+-- #align with_one.some_eq_coe WithOne.some_eq_coe
+
+@[simp, to_additive]
+theorem coe_ne_one {a : α} : (a : WithOne α) ≠ (1 : WithOne α) :=
+  Option.some_ne_none a
+#align with_one.coe_ne_one WithOne.coe_ne_one
+
+@[simp, to_additive]
+theorem one_ne_coe {a : α} : (1 : WithOne α) ≠ a :=
+  coe_ne_one.symm
+#align with_one.one_ne_coe WithOne.one_ne_coe
+
+@[to_additive]
+theorem ne_one_iff_exists {x : WithOne α} : x ≠ 1 ↔ ∃ a : α, ↑a = x :=
+  Option.ne_none_iff_exists
+#align with_one.ne_one_iff_exists WithOne.ne_one_iff_exists
+
+-- porting note : waiting for `lift` tactic
+--@[to_additive]
+--instance canLift : CanLift (WithOne α) α coe fun a => a ≠ 1 where prf a := ne_one_iff_exists.1
+--#align with_one.can_lift WithOne.canLift
+
+@[simp, norm_cast, to_additive]
+theorem coe_inj {a b : α} : (a : WithOne α) = b ↔ a = b :=
+  Option.some_inj
+#align with_one.coe_inj WithOne.coe_inj
+
+-- port note: at the time of writing it seems that `@[norm_cast, to_additive]` doesn't
+-- put the norm_cast tag on the to_additivised declaration. This will cause problems
+-- with `norm_cast` failing on autoported proofs.
+attribute [norm_cast] WithZero.coe_inj
+
+@[elab_as_elim, to_additive]
+protected theorem cases_on {P : WithOne α → Prop} : ∀ x : WithOne α, P 1 → (∀ a : α, P a) → P x :=
+  Option.casesOn
+#align with_one.cases_on WithOne.cases_on
+
+-- porting note: in Lean 3 there was the following comment:
+-- the `show` statements in the proofs are important, because otherwise the generated lemmas
+-- `with_one.mul_one_class._proof_{1,2}` have an ill-typed statement after `with_one` is made
+-- irreducible.
+@[to_additive]
+instance [Mul α] : MulOneClass (WithOne α) where
+  mul := (· * ·)
+  one := 1
+  one_mul := show ∀ x : WithOne α, 1 * x = x from (Option.liftOrGet_isLeftId _).1
+  mul_one := show ∀ x : WithOne α, x * 1 = x from (Option.liftOrGet_isRightId _).1
+
+@[to_additive]
+instance [Semigroup α] : Monoid (WithOne α) :=
+  { instMulOneClassWithOne with mul_assoc := (Option.liftOrGet_isAssociative _).1 }
+
+-- example [Semigroup α] :
+--     @Monoid.toMulOneClass _ (@instMonoidWithOne α _) = @instMulOneClassWithOne α _ :=
+--   rfl
+
```

Comment:
Delete?

Reply:
I'd have to make a corresponding PR to mathlib3 then, right? https://github.com/leanprover-community/mathlib/blob/c982179ec21091d3e102d8a5d9f5fe06c8fafb73/src/algebra/group/with_one/defs.lean#L127-L128